compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -10287,6 +10356,98 @@ private STNode parseLetClause(boolean isRhsExpr) {
         return STNodeFactory.createLetClauseNode(letKeyword, letVarDeclarations);
     }
 
+    /**
+     * Parse order by clause.
+     * <p>
+     * <code>order-by-clause := order by expression (ascending | descending)? (expression (ascending | descending)?)*
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOrderByClause(boolean isRhsExpr) {
+        STToken nextToken = peek();
+        if (nextToken.kind != SyntaxKind.ORDER_KEYWORD) {
+            return STNodeFactory.createEmptyNode();
+        }
+        STNode orderKeyword = parseOrderKeyword();
+        STNode byKeyword = parseByKeyword();
+        STNode orderKeys = parseOrderKeyList(ParserRuleContext.ORDER_KEY, isRhsExpr);
+
+        byKeyword = cloneWithDiagnosticIfListEmpty(orderKeys, byKeyword,
+                DiagnosticErrorCode.ERROR_MISSING_EXPRESSION);
+
+        return STNodeFactory.createOrderByClauseNode(orderKeyword, byKeyword, orderKeys);
+    }
+
+    /**
+     * Parse order key.
+     * <p>
+     * <code>order-key-list := order-key [, order-key]*</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOrderKeyList(ParserRuleContext context, boolean isRhsExpr) {
+        startContext(context);
+        List<STNode> orderKeys = new ArrayList<>();
+        STToken nextToken = peek();
+
+        // Make sure at least one order key is present
+        if (isEndOfOrderKeys(nextToken.kind)) {
+            endContext();
+            return STNodeFactory.createEmptyNodeList();
+        }
+
+        // Parse first order key, that has no leading comma
+        STNode orderKey = parseOrderKey(isRhsExpr);
+        orderKeys.add(orderKey);
+
+        // Parse the remaining order keys
+        nextToken = peek();
+        STNode leadingComma;
+        while (!isEndOfOrderKeys(nextToken.kind)) {
+            leadingComma = parseComma();
```

Comment:
When a comma separated list is parsed, theres a possibility of having some extraneous token after a member, that is not captured by `isEndOfOrderKeys`. 

Therefore we check the token before parsing the separator - because after recovering, we could even reach the end of the list. So we have to do something similar to `parseArgEnd()`/`parseListConstructorMemberEnd()` methods.